src/main/engine/engine.cpp

```diff
@@ -93,16 +112,19 @@ void Engine::render( vector<unique_ptr<GameComponent>>& components )
   should_update = true;
 }
 
-void Engine::update( vector<unique_ptr<GameComponent>>& components )
+void Engine::update_components( InputEvent& input_event )
 {
   for( auto& component : components )
   {
-    if( frame_count % component -> get_frames_per_update() == 0 && should_update )
+    if( should_update )
```

Comment:
Using states for the falling piece cleans up the engine code nicely. Ideally, the engine is ignorant of almost everything about its components, and just calls `update()` when it needs to.